### PR TITLE
Event horizon metrics

### PR DESCRIPTION
**Pull request by woksin**
## Summary

Adds metrics for EventHorizon

### Added

- Metrics

### CODE REVIEW

File: Source/EventHorizon/Consumer/Connections/MetricsCollector.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            _totalSuccessfulResponses = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_consumer_connections_successful_responses_total",
+                "EventHorizonConnection total number of successful connection responses");
+
+            _totalFailureResponses = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_consumer_connections_failure_responses_total",
+                "EventHorizonConnection total number of failure connection responses");
+
+            _totalEventHorizonEventsHandled = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_consumer_connections_event_horizon_events_handled_total",
```
Comment (jakhog) [inline Source/EventHorizon/Consumer/Connections/MetricsCollector.cs:45]:
```suggestion
                "dolittle_shared_runtime_event_horizon_consumer_handled_events_total",
```

File: Source/EventHorizon/Producer/MetricsCollector.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        Counter _totalInncommingSubscriptions;
+        Counter _totalRejectedSubscriptions;
+        Counter _totalAcceptedSubscriptions;
+        Counter _totalEventsWrittenToEventHorizon;
+
+        /// <inheritdoc/>
+        public IEnumerable<Collector> Provide(IMetricFactory metricFactory)
+        {
+            _totalInncommingSubscriptions = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_producer_consumer_service_inncomming_subscriptions_total",
```
Comment (jakhog) [inline Source/EventHorizon/Producer/MetricsCollector.cs:26]:
```suggestion
                "dolittle_shared_runtime_event_horizon_producer_incoming_subscriptions_total",
```

File: Source/EventHorizon/Producer/MetricsCollector.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                "ConsumerService total number of subscription received from other Runtimes");
+
+            _totalRejectedSubscriptions = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_producer_consumer_service_rejected_subscriptions_total",
```
Comment (jakhog) [inline Source/EventHorizon/Producer/MetricsCollector.cs:30]:
```suggestion
                "dolittle_shared_runtime_event_horizon_producer_rejected_subscriptions_total",
```

File: Source/EventHorizon/Producer/MetricsCollector.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                "ConsumerService total number of rejected subscriptions");
+
+            _totalAcceptedSubscriptions = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_producer_consumer_service_accepted_subscriptions_total",
```
Comment (jakhog) [inline Source/EventHorizon/Producer/MetricsCollector.cs:34]:
```suggestion
                "dolittle_shared_runtime_event_horizon_producer_accepted_subscriptions_total",
```

File: Source/EventHorizon/Producer/MetricsCollector.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                "ConsumerService total number of accepted subscriptions");
+            _totalEventsWrittenToEventHorizon = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_producer_consumer_service_event_horizon_events_written_total",
```
Comment (jakhog) [inline Source/EventHorizon/Producer/MetricsCollector.cs:37]:
```suggestion
                "dolittle_shared_runtime_event_horizon_producer_events_written_total",
```

File: Source/EventHorizon/Consumer/MetricsCollector.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        /// <inheritdoc/>
+        public void IncrementTotalRegisteredSubscriptions()
+            => _totalRegisteredSubscriptions.Inc();
+
+        /// <inheritdoc/>
+        public void IncrementTotalConnectedSubscriptions()
+            => _totalConnectedSubscriptions.Inc();
+
+        public void DecrementTotalConnectedSubscriptions()
+            => _totalConnectedSubscriptions.Dec();
```
Comment (jakhog) [inline Source/EventHorizon/Consumer/MetricsCollector.cs:58]: If it is a Gauge - it is not a total number (maybe current?)

File: Source/EventHorizon/Consumer/MetricsCollector.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            _totalSubscriptionRequests = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_consumer_subscription_requests_total",
+                "Subscriptions total number of subscription requests received from Head");
+
+            _totalRegisteredSubscriptions = metricFactory.Counter(
+                "dolittle_shared_runtime_event_horizon_consumer_registered_subscriptions_total",
+                "Subscriptions total number of registered subscriptions");
+
+            _totalConnectedSubscriptions = metricFactory.Gauge(
+                "dolittle_shared_runtime_event_horizon_consumer_connected_subscriptions_total",
```
Comment (jakhog) [inline Source/EventHorizon/Consumer/MetricsCollector.cs:34]: If it is a Gauge - it is not a total.
```suggestion
                "dolittle_shared_runtime_event_horizon_consumer_connected_subscriptions",
```

File: Source/EventHorizon/Consumer/Connections/EventHorizonConnection.cs [hunk 8]
```diff
@@ -44,20 +48,23 @@ public class EventHorizonConnection : IEventHorizonConnection
         {
             if (await _reverseCallClient.Connect(CreateRequest(subscription, publicEventsPosition), cancellationToken).ConfigureAwait(false))
             {
+                _metrics.IncrementTotalConnectionAttempts();
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/EventHorizon/Consumer/Connections/EventHorizonConnection.cs:51]: So the timespan between calling connect and it actually connecting?

File: Source/EventHorizon/Producer/ConsumerService.cs [hunk 9]
```diff
@@ -112,7 +116,7 @@ public void Dispose()
             var (dispatcher, arguments) = tryConnect.Result;
             _executionContextManager.CurrentFor(arguments.ExecutionContext);
 
-
+            _metrics.IncrementTotalInncommingSubscriptions();
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/EventHorizon/Producer/ConsumerService.cs:119]: My spellchecker extension complained about that for some reason. I don't know who's right :D

File: Source/EventHorizon/Consumer/MetricsCollector.cs (hunk 6 above)
Comment (woksin) [reply Source/EventHorizon/Consumer/MetricsCollector.cs:58]: Yeah, that's true

**Review by joelhoisko — APPROVED** [review]
Looks good, I changed the metrics bootup log to information so that it would pop up when starting the Runtime. I also updated `prometheus-net` to 4.1.1 (latest) and fixed some ambiguous naming references it bought with it.

**Pull request merged**